Add topic classification via post answer composition workflow

davidgisbey opened the pull request:
## Description

This spikes out implementing topic tagging for questions. This is run post answer composition because we need to run tagging on the rephrased question if the question was rephrased.

It currently points towards this [PR](https://github.com/alphagov/govuk_chat_private/pull/30) on the private repo which adds the required config for the LLM call. It should be noted that this is based on some info provided by David Gillespie and isn't the finalised version. I created the system prompt just to get something working, but the topics and tool call were largely provided by David. Def still subject to change though.

This PR:
- adds an AnswerAnalysis model and table to the db which will capture various metadata we want for analysis. For this it captures
  - primary topic
  - secondary topic
  - metics
  - llm respones
- adds an AnswerInsightsJob which is called at the end of answer composition
- adds the AnswerInsight::TopicTagger which  is called within the above job. This makes the LLM call to tag topics to the question and persists an AnswerAnalysis on the recently composed anser
- adds the topics, metrics and llm responses to the admin question show page
- adds filter to the questions index page to filter on primary and secondary topic

It should be noted that we only want to tag topics for answers which successfully completed the pipeline and weren't triggered by malicious user behaviour (jailbreak guardrails). So these have been filtered out 

## Screenshots 
### Question index

<img width="746" height="720" alt="image" src="https://github.com/user-attachments/assets/d85c9a99-5289-4860-801b-6802c578f011" />

<img width="611" height="781" alt="image" src="https://github.com/user-attachments/assets/98fe0106-cdec-4bff-83b9-cdd8498b66ef" />


### Question show 

<img width="611" height="517" alt="image" src="https://github.com/user-attachments/assets/6b8d1f61-da9c-4ca1-8776-42995f94be47" />

<img width="594" height="800" alt="image" src="https://github.com/user-attachments/assets/505cdbe3-2e75-4153-9749-476ddd7716ce" />

<img width="588" height="492" alt="image" src="https://github.com/user-attachments/assets/bca39122-e605-42bc-8585-f1c2305314ce" />

davidgisbey commented on config/answer_topic.yml:
```diff
@@ -0,0 +1,154 @@
+system_prompt: |
```
obviously this file will live in the private repo, but i've just added it here for now so it's easy to see/review in one place.

kevindew commented on app/jobs/compose_answer_job.rb:
```diff
@@ -10,6 +10,7 @@ def perform(question_id)
 
     begin
       answer.save!
+      AnswerTopic::Tagger.call(Answer.includes(%i[topic question]).find(answer.id))
```
I think we probably want this outside the begin/rescue since the rescue isn't really for this (and would be very confusing if it triggered).

I imagine we also want this to run against a question that has timed out? it could even be that we want to change the way guardclauses are done so it doesn't exit early on question.answer but instead checks for question.answer, if none generates an answer, then answer.topic, if non tag a topic.

davidgisbey replied:
yea the reason i haven't implemented the error timeout one is i'm not sure we would want to run it on a canned answer? So i've held off doing anymore until i get a clear answer which answer statuses we want to add topics too. But i'll def keep it in mind.

Good shout about moving it out. I did end have 

```
AnswerTopic::Tagger.call(Answer.includes(%i[topic question]).find(answer.id)) if answer.persisted?
```

below but moved it up. Should've left it

kevindew commented on lib/answer_topic/tagger.rb:
```diff
@@ -0,0 +1,14 @@
+class AnswerTopic::Tagger
```
So I think we should avoid using AnswerTopic as the root namespace here since it's already a model and thus can't be a module.

Perhaps AnswerAnalysis would be a good route namespace and this could be TopicTagger ?

davidgisbey replied:
good shout will do

kevindew commented on lib/answer_topic/tagger.rb:
```diff
@@ -0,0 +1,14 @@
+class AnswerTopic::Tagger
+  def self.call(answer)
+    answer_strategy = answer.question.answer_strategy.to_sym
```
As discussed verbally, it'd be good if this didn't need to be aware of answer strategy - we'll just use claude as the tagger until we need to consider otherwise.

I imagine this may have a knock on that we might need to set-up heroku OpenAI env to have AWS creds if it doesn't already.

davidgisbey replied:
Good shout. I'll make sure i change this post chat with David. (or today if i have time after the BigQuery stuff.)

kevindew commented on lib/answer_topic/claude/tagger.rb:
```diff
@@ -0,0 +1,90 @@
+
+  def inference_config
+    {
+      max_tokens: 4096,
```
Not something we'd need to change right now, but we will want this to likely be quite a lot smaller for this task

davidgisbey replied:
Yea agreed. I just added this as a placeholder until David feeds back on his work

kevindew reviewed (commented):
Looks really good, just a few implementation comments. Happy to discuss more.

As an aside, I do wonder if we might be reproducing the same sort of Claude call a model, log metrics and llm_response a few times. Not sure if there's a potentially helper class or function potentially looming.

davidgisbey commented on app/jobs/answer_insights_job.rb:
```diff
@@ -0,0 +1,10 @@
+class AnswerInsightsJob < ApplicationJob
```
Yea sounds good. I added this prior to knowing whether we were going with an async approach or not, but given we're going to call multiple jobs concurrently i think we can change this now

davidgisbey commented on lib/answer_insights/topic_tagger.rb:
```diff
@@ -0,0 +1,96 @@
+class AnswerInsights::TopicTagger
```
Yea i was trying to find a name that covers generating analysis and manual eval work, but it might be that they have sep models. 

I have a preference for `AnswerAnalysisGeneration` out of the two. `AnswerAnalysing` feels a bit clunkier

davidgisbey commented on app/views/admin/questions/_question_filters.html.erb:
```diff
@@ -179,6 +179,36 @@
+    label: "Primary topic",
+    heading_size: "s",
+    full_width: true,
+    options: [{text: "", value: ""}] + Rails.configuration.govuk_chat_private.llm_prompts.claude.topic_tagger["topics"].keys.map do |topic|
```
Yea it's an interesting one. We don't use answer_statuses at all in the prompts. All the logic for it is in the public codebase so that's why i leaned towards the private repo. Having said that i don't think it's something we particularly case about being public.

I guess what we need to consider is. Do we want all info used in the prompt to be private? In which case it should live in `govuk_chat_private.analysis_topics`. If not then we can just add a config file in this repo.

I'm leaning towards the private repo. Just becuase it feels like it's easier to reason what should live where when  anything directly used in prompts lives there. 

Otherwise i feel like each time we add something new like this we will need to discuss where it lives. Wdyt?

kevindew replied:
Yeah a better example I should have picked is question_routing_labels.yml as something where we have bits in both places - so probably a better analogue.

Yeah we could go all for the private repo until we see evidence of a problem, at least that means changes could be atomic.

davidgisbey replied:
ah right i see what you mean. Maybe it should just live in the public one then if there's prior art of it being done. It doesn't really feel like it needs to live in the private repo and we do use it several times in the admin UI.

And i can't imagine any added security concerns considering we're not running it on Jailbreaks.

davidgisbey commented:
@kevindew i ended up adding the topics into config in this repo. I've left a couple of new things in the last few commits:

1. i've locked the answer before creating/updating the answer analysis in the 3rd to last commit
2. i've moved metric and llm_response assignment to a shared concern
3. i've added retry logic to the topics job

davidgisbey commented on spec/models/answer_analysis_spec.rb:
```diff
@@ -41,4 +41,37 @@
     end
   end
+
+  describe "#assign_metrics" do
```
these can probs live in a shared example, but for now i just did this until i got some feedback on what people thought about the abstraction.

kevindew commented on lib/answer_analysis_generation/topic_tagger.rb:
```diff
@@ -0,0 +1,100 @@
+    )
+
+    answer.reload.with_lock do
+      raise "Topics already generated for answer #{answer.id}" if answer.analysis&.primary_topic.present?
```
Is this exception likely at all? I'm thinking it's probably super rare, i.e. could only be caused by somehow sidekiq getting duplicate jobs and running them concurrently, so probably fine. But did wonder if there was a likely scenario for it we'd not want it to polute our error logs and would potentially want it as something caught by the job.

davidgisbey replied:
Yea it really shouldn't occur. Def happy to take it out.

kevindew replied:
Fine (and probably better) to keep in if it's a last ditch state check. 

Just wanted to check that we should have something that is less catasrophic if the scenario is actually likely.

kevindew reviewed (commented):
I've run out of time so only a quick review but think comments should be interesting at least

davidgisbey commented on lib/answer_analysis_generation/topic_tagger.rb:
```diff
@@ -0,0 +1,100 @@
+      **inference_config,
+    )
+
+    answer.reload.with_lock do
```
So the reason i locked the answer here which might be premature is that soon we're going to be running multiple jobs with llm calls concurrently

1. topics
2. question intent
3. question category (general, specific etc.)

Each of these could finish before the other 2 and would then need to either

a. create the analysis
b. update the existing analysis

so i figured locking the answer would be the simplest solution. I guess we could do 

```
 analysis = answer.analysis || answer.build_analysis
```

and catch all errors for retries so if `raise "Analysis already created for answer #{answer.id}"` is raised then it'll just get retried.

kevindew replied:
So I expect we'd do something different if we have multiple jobs. What I expect that would look like is:

```
# find_or_create_by! handles a RecordNotUnique exception
analysis = AnswerAnalysis.find_or_create_by!(answer_id: answer.id)

# do the analysis

analysis.with_lock do
  # any guard clauses that this analysis already exists
  analysis.update!(# however it has changed)
end
```

davidgisbey replied:
Okay cool. I went with your first suggestion for now since we can move to the second when the need arises.

davidgisbey commented on app/jobs/answer_topics_job.rb:
```diff
@@ -1,5 +1,7 @@
 class AnswerTopicsJob < ApplicationJob
+  MAX_RETRIES = 5
   queue_as :default
+  retry_on Anthropic::Errors::APIError, wait: 1.minute, attempts: MAX_RETRIES
```
Ah yea good point. 
``` pry(main)> AnswerTopicsJob.sidekiq_options
=> {"retry" => true, "queue" => "default"}
```
so we can either toggle them off in the job via `sidekiq_options retry: false` or we could just retry via Sidekiq. Do you have a preference?

kevindew replied:
I think turning them off at sidekiq is better to avoid two systems. The fact there are two systems is a bit of a source of pain (the ActiveJob retries won't end up in the Sidekiq retry queue but in the scheduled queue)

exfalsoquodlibet commented on db/schema.rb:
```diff
@@ -23,6 +23,17 @@
   create_enum "guardrails_status", ["pass", "fail", "error"]
   create_enum "question_routing_label", ["about_mps", "advice_opinions_predictions", "character_fun", "genuine_rag", "gov_transparency", "greetings", "harmful_vulgar_controversy", "multi_questions", "negative_acknowledgement", "non_english", "personal_info", "positive_acknowledgement", "vague_acronym_grammar", "unclear_intent", "requires_account_data"]
 
+  create_table "answer_analyses", id: :uuid, default: -> { "gen_random_uuid()" }, force: :cascade do |t|
+    t.string "primary_topic"
+    t.string "secondary_topic"
+    t.jsonb "metrics"
+    t.jsonb "llm_responses"
+    t.uuid "answer_id", null: false
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.index ["answer_id"], name: "index_answer_analyses_on_answer_id", unique: true
```
Maybe I am missing it as my understanding of Ruby is limited, but could we also record the question that was actually tagged?

davidgisbey replied:
So it's associated with the answer, since we pass the answer into the tagger once it's persisted in the database so we have access to the rephrased question (which lives on the answer).

But the answer is associated with the question. So we can just do `analyisis.answer.question` to get access to the question that was tagged.

exfalsoquodlibet replied:
Thanks, David, I was tied up in back-to-back meetings.

I had a quick look at the DB schema and asked Claude to represent it in a diagram. I can see that rich metadata is stored using JSONB for `metrics` and `llm_response`, which makes sense since these fields need to be flexible enough to handle evolving data structures.

Just to confirm my understanding:

* **`metrics`** captures something like:

  ```ruby
  def build_metrics(response, start_time)
    {
      duration: Clock.monotonic_time - start_time,
      llm_prompt_tokens: BedrockModels.claude_total_prompt_tokens(response[:usage]),
      llm_completion_tokens: response[:usage][:output_tokens],
      llm_cached_tokens: response[:usage][:cache_read_input_tokens],
      model: response[:model],
    }
  end
  ```

* **`llm_response`** stores the raw JSON response from the LLM call (e.g. from the TopicTagger), following a structure like:

  ```ruby
  expected_llm_response = {
    "id" => "msg-id",
    "content" => [
      {
        "id" => "tool-use-id",
        "input" => {
          "primary_topic" => "business",
          "secondary_topic" => "benefits",
          "confidence" => "high",
          "reasoning" => "reason"
        },
        "name" => "tagger_reasoning",
        "type" => "tool_use"
      }
    ],
    "model" => BedrockModels::CLAUDE_SONNET,
    "role" => "assistant",
    "stop_reason" => "tool_use",
    "type" => "message",
    "usage" => {
      "cache_read_input_tokens" => 20,
      "input_tokens" => 10,
      "output_tokens" => 20
    },
  }
  ```

This is just me trying to understand exactly what we’re storing and how, nothing that should block this PR from merging.

kevindew commented:
Maybe we can grab some time (after standup perhaps) to talk about naming?

kevindew commented on app/helpers/admin/questions_helper.rb:
```diff
@@ -201,6 +201,20 @@ def question_show_summary_list_rows(question, answer, question_number, total_que
+          value: answer.analysis.primary_topic.humanize,
+        }
+
+        if answer.analysis.secondary_topic.present?
```
I think we should probably show this as empty if we have a primary topic. 

On the basis that this not actually having a result is still an outcome of analysis.

kevindew commented on app/jobs/answer_topics_job.rb:
```diff
@@ -0,0 +1,17 @@
+class AnswerTopicsJob < ApplicationJob
+  MAX_RETRIES = 5
+  queue_as :default
```
I think we might have some lingering work to do on prioritising job execution - so this can't block answer composition. We can do this post merge though

davidgisbey replied:
Sounds good. I'll card something up for it once this has been merged.

kevindew commented on app/models/answer_analysis.rb:
```diff
@@ -0,0 +1,13 @@
+  }
+
+  def serialize_for_export
+    as_json.merge("llm_responses" => llm_responses.to_json)
```
A micro-optimisation is to exclude llm_responses from as_json so you don't have the penalty of that being serialised for no reason.

kevindew commented on config/application.rb:
```diff
@@ -66,6 +66,7 @@ class Application < Rails::Application
     config.answer_statuses = Hashie::Mash.new(YAML.load_file("#{__dir__}/answer_statuses.yml"))
     config.question_routing_labels = Hashie::Mash.new(YAML.load_file("#{__dir__}/question_routing_labels.yml"))
     config.search = Hashie::Mash.new(YAML.load_file("#{__dir__}/search.yml"))
+    config.topics = YAML.load_file("#{__dir__}/topics.yml")
```
Could we call this question_topics - just to get a prefix in there early as a domain concept to avoid clashes with any other type of topic. 

It doesn't necessary to change this on the Analysis model as we already have the context of it being on answer

kevindew replied:
Also, I know I'm treading old ground but just wondering given we're using the list from the prompt at the moment whether we should find a way to avoid having two lists to maintain for now (given there's not any deviation) 

I think perhaps the way that seems least painful is if we populated this config variable with data from the private gem (I think dependencies are available here)? Could even have a test that asserts it's a non empty array. That strikes me as maybe a way to avoid duplication but not have deeply mixed domain concepts.

davidgisbey replied:
Okay great. I've used the topics in input_schema of the private repo and just added a basic test to check that there's an array containing something when you pulll the schema configs enum.

davidgisbey commented on app/jobs/answer_topics_job.rb:
```diff
@@ -0,0 +1,17 @@
+      return logger.info("Answer #{answer_id} is not eligible for topic analysis")
+    end
+
+    AnswerAnalysisGeneration::TopicTagger.call(answer)
```
yup makes sense 👍 i guess down the line we just need to find_or_init and lock the instance.

davidgisbey commented:
@kevindew thanks for the review. Just getting to look at it now as i working on gradio stuff today, but i'll prioritise getting this done next

davidgisbey commented on app/views/admin/questions/_question_filters.html.erb:
```diff
@@ -1,6 +1,11 @@
+                         .govuk_chat_private
+                         .llm_prompts
+                         .claude
+                         .topic_tagger.dig("tool_spec", "input_schema", "$defs", "govuk_topic_tags", "enum")
```
So it doesn't have access to `config.govuk_chat_private` here, but i can do 

```
    config.question_topics = GovukChatPrivate.config
                                                                              .llm_prompts.claude
                                                                              .topic_tagger
                                                                             .dig("tool_spec", "input_schema", "$defs", "govuk_topic_tags", "enum")
                                                                              .sort

```

or do it in an intialiser. Do you have a preference?

davidgisbey commented:
@kevindew  i've added some stuff to a TODO list

kevindew commented on spec/models/answer_analysis_spec.rb:
```diff
@@ -0,0 +1,48 @@
+  end
+
+  describe ".exportable" do
+    let(:new_answer_analysis) { create(:answer_analysis, created_at: 2.days.ago) }
+    let(:old_answer_analysis) { create(:answer_analysis, created_at: 4.days.ago) }
+
+    before do
+      new_answer_analysis
+      old_answer_analysis
+    end
```
Ah sure - that warning makes sense. I'd say the most pragmatic approach for immediate way is to do:

```
let!(:exported_answer_analysis) { create(:answer_analysis, created_at: 2.days.ago) }
before { create(:answer_analysis, created_at: 4.days.ago) }
```

I imagine if we refactor things more for a shared we may want to explicitly assert the item that isn't there. 

e.g. 

```
expect(something).to include(exportable_answer_analysis)
expect(something).not_to include(exportable_answer_analysis)
```

but doesn't feel right for this exact test

davidgisbey replied:
okay cool. have sorted that

kevindew commented:
Great - think we're just about there on this now

davidgisbey commented:
Thanks @kevindew i've made that final change. I'm now working on the follow up tasks. I'll card them up now.
